### PR TITLE
xonsh: Add wrapper

### PR DESCRIPTION
**Pull request by adisbladis**
###### Description of changes

It is currently not obvious how to install/use xonsh with dependencies and python packages.

This PR implements a wrapper that allows you to construct a custom xonsh environment by using:
``` nix
xonsh.override { extraPackages = ps: [ ps.requests ]; }
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/xonsh.nix [hunk 1]
```diff
@@ -28,7 +28,7 @@ in
         type = types.package;
         default = pkgs.xonsh;
         defaultText = literalExpression "pkgs.xonsh";
-        example = literalExpression "pkgs.xonsh.override { configFile = \"/path/to/xonshrc\"; }";
+        example = literalExpression "pkgs.xonsh.override { extraPackages = ps: [ ps.requests ]; }";
```
Comment (SuperSandro2000) [inline nixos/modules/programs/xonsh.nix:31]:
```suggestion
        example = literalExpression "pkgs.xonsh.override { extraPackages = ps: with ps; [ requests ]; }";
```

File: pkgs/shells/xonsh/wrapper.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+in
+runCommand "${xonsh.pname}-${xonsh.version}"
+{
+  inherit (xonsh) pname version meta;
+} ''
+  mkdir -p $out/bin
+  for bin in ${lib.getBin xonsh}/bin/*; do
+    ln -s ${pythonEnv}/bin/$(basename "$bin") $out/bin/
+  done
```
Comment (SuperSandro2000) [inline pkgs/shells/xonsh/wrapper.nix:23]:
```suggestion
  lndir ${lib.getBin xonsh}/bin $out/bin
```
lndir is under xorg.lndir but not important just an idea

**Review by SuperSandro2000 — APPROVED** [review]

**Review by jian-lin — APPROVED** [review]

File: pkgs/shells/xonsh/default.nix [hunk 3]
```diff
@@ -90,5 +77,6 @@ python3Packages.buildPythonApplication rec {
 
   passthru = {
     shellPath = "/bin/xonsh";
```
Comment (SamLukeYes) [inline pkgs/shells/xonsh/default.nix:79]: My NixOS configuration (https://github.com/SamLukeYes/nixos/commit/7bf704df6ea851ff68a3ad005f24178ee84e8e64) fails to build with this change.

```
error: /nix/store/0bq90g80r0b1s9s8ipznsh1azhsl8swp-xonsh-0.14.0 is not a shell package
```

Perhaps the wrapper package should pass through `shellPath`, I suppose.

Comment (adisbladis) [reply pkgs/shells/xonsh/default.nix:79]: It does inherit `passthru` so I'm not sure why you're seeing this.

That error comes from `nixos/lib/utils.nix`. I copy pasted from there to make this minimal case:
``` nix
let
  pkgs = import ./. { };
  inherit (pkgs) lib;
  inherit (lib) types;

  toShellPath = shell:
    if types.shellPackage.check shell then
      "/run/current-system/sw${shell.shellPath}"
    else if types.package.check shell then
      throw "${shell} is not a shell package"
    else
      shell;

in toShellPath pkgs.xonsh

```
which works.

Comment (SamLukeYes) [reply pkgs/shells/xonsh/default.nix:79]: > It does inherit `passthru` so I'm not sure why you're seeing this.

No, it doesn't.

```
nix-repl> legacyPackages.x86_64-linux.xonsh.passthru
{ }
```

> That error comes from `nixos/lib/utils.nix`. I copy pasted from there to make this minimal case:
> 
> ```nix
> let
>   pkgs = import ./. { };
>   inherit (pkgs) lib;
>   inherit (lib) types;
> 
>   toShellPath = shell:
>     if types.shellPackage.check shell then
>       "/run/current-system/sw${shell.shellPath}"
>     else if types.package.check shell then
>       throw "${shell} is not a shell package"
>     else
>       shell;
> 
> in toShellPath pkgs.xonsh
> ```
> 
> which works.

I also tried this on your branch. It throws the error.

**Review by SamLukeYes — APPROVED** [review]
Works for me now. Thanks!

File: pkgs/shells/xonsh/default.nix (hunk 3 above)
Comment (adisbladis) [reply pkgs/shells/xonsh/default.nix:79]: I forgot to push the change that I had made locally! Doh!

File: pkgs/shells/xonsh/wrapper.nix (hunk 2 above)
Comment (adisbladis) [reply pkgs/shells/xonsh/wrapper.nix:23]: If you look a bit closer you can see that these are not the same.
What I do in this wrapper is to link a subset of binaries (those available in the xonsh package) from a python environment.
This is a filtering mechanism.

**Pull request merged**